### PR TITLE
[GLM] fix bug where GLM model would have null _input_params

### PR DESCRIPTION
**Pull request by honzasterba**
Fixes multiple issues:
- computeParams was not locking and updating the Model instance
- GLM is not releasing lock on Model instance in computeImpl but only in onCompletion, makint the above impossible

Contains also removal of some unnecessary code.

### CODE REVIEW

File: h2o-algos/src/main/java/hex/glm/GLM.java [hunk 1]
```diff
@@ -897,38 +894,20 @@ private final double lmax(double[] grad) {
...
-      } catch(Throwable t){
-        // nada
       }
     }
+
     private transient Cholesky _chol;
     private transient L1Solver _lslvr;
 
-
-    int [] findZeros(double [] vals){
```
Comment (honzasterba) [inline h2o-algos/src/main/java/hex/glm/GLM.java:920]: unused

File: h2o-algos/src/main/java/hex/glm/GLM.java [hunk 2]
```diff
@@ -897,38 +894,20 @@ private final double lmax(double[] grad) {
...
-          }
+      if(_parms._lambda_search && _parms._is_cv_model)
+        Scope.untrack(removeLater(_dinfo.getWeightsVec()._key));
+      if (_parms._HGLM) {
+        Key[] vecKeys = _toRemove;
+        for (int index=0; index < vecKeys.length; index++) {
+          Vec tempVec = DKV.getGet(vecKeys[index]);
+          tempVec.remove();
         }
-        if(_model!=null)
```
Comment (honzasterba) [inline h2o-algos/src/main/java/hex/glm/GLM.java:910]: ulocking moved to compute method

File: h2o-algos/src/main/java/hex/glm/GLM.java [hunk 3]
```diff
@@ -897,38 +894,20 @@ private final double lmax(double[] grad) {
...
-          _model.unlock(_job);
-      } catch(Throwable t){
```
Comment (honzasterba) [inline h2o-algos/src/main/java/hex/glm/GLM.java:912]: this seem reckless, I think we want to know if something breaks during cleanup

File: h2o-algos/src/main/java/hex/glm/GLM.java [hunk 4]
```diff
@@ -2131,6 +2110,14 @@ protected Submodel computeSubmodel(int i,double lambda, double nullDevTrain, dou
     
     @Override
     public void computeImpl() {
+      try {
+        doCompute();
+      } finally {
+        _model.unlock(_job);
```
Comment (michalkurka) [inline h2o-algos/src/main/java/hex/glm/GLM.java:2116]: Can _model be null at this point?

Comment (honzasterba) [reply h2o-algos/src/main/java/hex/glm/GLM.java:2116]: no, it is assigned to in init() and never set to anything else

File: h2o-algos/src/main/java/hex/glm/GLM.java (hunk 2 above)
Comment (michalkurka) [reply h2o-algos/src/main/java/hex/glm/GLM.java:910]: I think you have a valid point, however, I think this code needs more error handling if this catch is removed, eg.:

Can _dinfo be null? Can the temp vec already be removed from DKV (line 903 possible NPE).

I propose this: 
1) Add the error handling 
2) in onCompletion - invoke doCleanup the regular way
3) in onExceptionalCompletion - we are in an undefined state - any clean-up attempt is by definition best effort only - I would keep the try-catch - and produce a warning in the log

File: h2o-algos/src/main/java/hex/glm/GLM.java (hunk 3 above)
Comment (michalkurka) [reply h2o-algos/src/main/java/hex/glm/GLM.java:912]: I think you have a valid point, however, I think this code needs more error handling if this catch is removed, eg.:

Can _dinfo be null? Can the temp vec already be removed from DKV (line 903 possible NPE).

I propose this:

Add the error handling
in onCompletion - invoke doCleanup the regular way
in onExceptionalCompletion - we are in an undefined state - any clean-up attempt is by definition best effort only - I would keep the try-catch - and produce a warning in the log

File: h2o-algos/src/main/java/hex/glm/GLM.java (hunk 2 above)
Comment (michalkurka) [reply h2o-algos/src/main/java/hex/glm/GLM.java:910]: sounds good, @sebhrusen can you also take a look?

File: h2o-algos/src/main/java/hex/glm/GLM.java (hunk 4 above)
Comment (honzasterba) [reply h2o-algos/src/main/java/hex/glm/GLM.java]: you caught me there, and so did the smoke tests, apperently yes, the `_model` can be null here, fixed

File: h2o-algos/src/main/java/hex/glm/GLM.java (hunk 2 above)
Comment (sebhrusen) [reply h2o-algos/src/main/java/hex/glm/GLM.java:910]: `doCleanUp` is called in `onExceptionalCompletion` (typically if job is aborted), so `_model` is always null in this case.

What would be cleaner probably is to do the unlock only in `onCompletion`, right after calling `doCleanUp`

Comment (sebhrusen) [reply h2o-algos/src/main/java/hex/glm/GLM.java:910]: having it moved after compute looks good to me

Comment (honzasterba) [reply h2o-algos/src/main/java/hex/glm/GLM.java:910]: actually no, all other algos do unlock in `computeImpl` and this is necessary for the fix in `computeParams` to work and be able to acquire the lock again

File: h2o-algos/src/main/java/hex/glm/GLM.java (hunk 3 above)
Comment (honzasterba) [reply h2o-algos/src/main/java/hex/glm/GLM.java:912]: you are correct, such a change is out of scope for this PR, hence I have reverted most of it and at least added some logging

Comment (sebhrusen) [reply h2o-algos/src/main/java/hex/glm/GLM.java:912]: it was `Throwable` really? at least it should be limited to `Exception`

File: h2o-algos/src/main/java/hex/glm/GLM.java (hunk 4 above)
Comment (wendycwong) [reply h2o-algos/src/main/java/hex/glm/GLM.java]: _model is instantiated right after init().  If something goes wrong in init(), _model can be null.

File: h2o-algos/src/main/java/hex/glm/GLM.java [hunk 5]
```diff
@@ -855,10 +855,7 @@ private GLMModel runGLMModel(boolean standardize, Family family, Link link, Key<
   private void buildModel() {
     _model = new GLMModel(_result, _parms, this, _state._ymu, _dinfo._adaptedFrame.lastVec().sigma(), _lmax, _nobs);
     _model._output.setLambdas(_parms);  // set lambda_min and lambda_max if lambda_search is enabled
-    
-    // clone2 so that I don't change instance which is in the DKV directly
-    // (clone2 also shallow clones _output)
-    _model.clone2().delete_and_lock(_job._key);
+    _model.delete_and_lock(_job);
   }
```
Comment (wendycwong) [inline h2o-algos/src/main/java/hex/glm/GLM.java:857]: What is the problem changing the instance in the DKV directly?  Is it going to be slower?  I am just curious why this is done by Tomas.

Comment (honzasterba) [reply h2o-algos/src/main/java/hex/glm/GLM.java:857]: it does not make any sense to me, thats why I am removing it

**Review by koniecsveta — APPROVED** [review]

**Review by wendycwong — APPROVED** [review]

**Pull request merged**